WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+
+When an event is created from JavaScript, the `dispatchSnapshot` will be
+automatically set based on the current `AsyncContext.Snapshot` where the event
+was allocated. A flag could be added to `EventInit` so the `dispatchSnapshot` is
+`null`, to allow polyfilling events without a dispatch context; although it
+could also be useful for the default behavior of that flag to depend on the
+particular `Event` subclass.
+
+An alternative approach would be to populate the `dispatchSnapshot` property
+when the event is dispatched, with the context in which it is dispatched. This
+is similar to the behavior of `currentTarget` or `eventPhase`. However, the
+dispatch context would stay the same across an event dispatch, whereas those
+properties will change. Furthermore, in most cases where events are created and
+dispatched from JavaScript, the creation and dispatch happen in the same
+context. So we don’t recommend this approach.
+
+### Design principles for dispatch snapshots
+
+In some cases, a single event might have multiple async dispatch contexts as its
+possible causes, because the incoming data flow for that event might have
+multiple branches that go back to different script executions. This is
+particularly important when the data flow in implementations is expected to be
+different from the spec in relevant ways.
+
+An example of this is HTML media playback events, where if the user clicks play,
+and in short succession JS code calls `videoEl.load()` and then `videoEl.play()`
+in two different contexts, all three data flow branches will merge, resulting in
+a single `load` event being fired.
+
+If some of those sources are browser-originated and some originate from JS, only
+the latter ones should be considered. Beyond that, each such individual event
+would have to be considered. For media playback events, this merge is caused by
+debouncing (i.e. if a load is already in progress, calling a method that would
+start one will reuse the existing one), and so the dispatch context should be
+that for the earliest web API call that resulted in this event. But other cases
+might have other needs, and their specifics need to be considered.
+
+## Runtime script errors
+
+The `error` event on a window or worker global object is fired whenever a script
+execution throws an uncaught exception. Having access to the context in which
+the exception was thrown can help determine which of multiple independent
+streams of execution triggered this error, and therefore how to clean up after
+it. Therefore,
+[`ErrorEvent`](https://html.spec.whatwg.org/multipage/webappapis.html#errorevent)
+will have a `throwSnapshot` property, reflecting the context in which the
+exception was thrown.
+
+
+## Promises / async web APIs
+
+The `unhandledrejection` event on a window or worker global object is fired
+whenever a promise resolves with a rejection, without having a rejection handler
+registered. Exposing the context in which the promise was rejected is useful for
```

Comment:
Maybe the complexity here is an argument to go back to ALS's semantics of using Promise allocation time instead. I don't think we really identified use cases where the semantics matter one way or another, and as Andreu pointed out elsewhere, for platform-vended promises, the answer will be the same as if it were allocation time.

Reply:
No, I also don't see cases where it would actually matter.  But I also still don't know what this document is actually saying - am I correct in understanding that the current text says it should be 2?

Reply:
I *think* it would be 2, yes (in which case it does coincide!) because the rejection happens during the `Promise.resolve`.